File: backend/src/main/java/lan/dk/podcastserver/manager/downloader/YoutubeDownloader.java
package lan.dk.podcastserver.manager.downloader;

import com.github.axet.vget.VGet;
import com.github.axet.vget.info.VGetParser;
import com.github.axet.vget.info.VideoFileInfo;
import com.github.axet.vget.info.VideoInfo;
import com.github.axet.wget.info.DownloadInfo;
import com.github.axet.wget.info.URLInfo;
import com.github.axet.wget.info.ex.DownloadError;
import com.github.axet.wget.info.ex.DownloadIOCodeError;
import com.github.axet.wget.info.ex.DownloadInterruptedError;
import com.github.axet.wget.info.ex.DownloadMultipartError;
import com.github.davinkevin.podcastserver.service.MimeTypeService;
import io.vavr.control.Try;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Status;
import lan.dk.podcastserver.repository.ItemRepository;
import lan.dk.podcastserver.repository.PodcastRepository;
import com.github.davinkevin.podcastserver.service.FfmpegService;
import com.github.davinkevin.podcastserver.service.factory.WGetFactory;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.springframework.context.annotation.Scope;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Objects;

import static io.vavr.API.Option;
import static io.vavr.API.Try;
import static java.lang.String.format;
import static java.lang.String.valueOf;
import static java.time.ZonedDateTime.now;
import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;

/**
 * Created by kevin on 14/12/2013 for Podcast Server
 */

@Slf4j
@Scope("prototype")
@Component("YoutubeDownloader")
public class YoutubeDownloader extends AbstractDownloader {

    private static final String DEFAULT_EXTENSION_MP4 = "mp4";
    private static final String ERROR_NO_CONTENT_TYPE = "Content Type %s not found for video %s at url %s";
    VGet v = null;

    final YoutubeWatcher watcher = new YoutubeWatcher(this);

    private final WGetFactory wGetFactory;
    private final FfmpegService ffmpegService;

    public YoutubeDownloader(ItemRepository itemRepository, PodcastRepository podcastRepository, PodcastServerParameters podcastServerParameters, SimpMessagingTemplate template, MimeTypeService mimeTypeService, WGetFactory wGetFactory, FfmpegService ffmpegService) {
        super(itemRepository, podcastRepository, podcastServerParameters, template, mimeTypeService);
        this.wGetFactory = wGetFactory;
        this.ffmpegService = ffmpegService;
    }

    @Override
    public Item download() {
        try {
            VGetParser parser = wGetFactory.parser(item.getUrl());
            v = wGetFactory.newVGet(parser.info(new URL(item.getUrl())));

            v.extract(parser, stopDownloading, watcher);

            target = Option(getTargetFile(item)).getOrElseThrow(() -> new RuntimeException("target is undefined for download of " + item.getUrl()));
            v
                .getVideo()
                .getInfo()
                .forEach(vi -> vi.setTarget(generatePartFile(target, vi).toFile()));

            v.download(parser, stopDownloading, watcher);
        } catch (DownloadInterruptedError e) {
            log.debug("Stopped", e);
        } catch (DownloadMultipartError e) {
            e.getInfo()
                .getParts()
                .stream()
                .map(DownloadInfo.Part::getException)
                .filter(Objects::nonNull)
                .forEach(Throwable::printStackTrace);

            throw new RuntimeException(e);
        } catch (StringIndexOutOfBoundsException | MalformedURLException | NullPointerException | DownloadError e) {
            throw new RuntimeException(e);
        }

        log.debug("Download ended");
        return item;  //To change body of implemented methods use File | Settings | File Templates.
    }

    private Path generatePartFile(Path targetFile, VideoFileInfo vi) {
        return targetFile.resolveSibling(targetFile.getFileName() + v.getContentExt(vi));
    }

    @Override
    public String getFileName(Item item) {
        return Option(v)
                .map(VGet::getVideo)
                .map(VideoInfo::getTitle)
                .map(s -> s.replaceAll("[^a-zA-Z0-9.-]", "_"))
                .getOrElseThrow(() -> new RuntimeException("Error during creation of filename of " + item.getUrl()));
    }

    @Override
    public void pauseDownload() {
        item.setStatus(Status.PAUSED);
        saveSyncWithPodcast();
        convertAndSaveBroadcast();
    }

    @Override
    public void restartDownload() {
        item.setStatus(Status.STARTED);
        saveSyncWithPodcast();
        convertAndSaveBroadcast();
        synchronized (watcher) { watcher.notifyAll(); }
    }

    @Override
    public void stopDownload() {
        if (item.getStatus() == Status.PAUSED) {
            synchronized (watcher) { watcher.notify(); }
        }
        super.stopDownload();

        if (nonNull(v) && nonNull(v.getVideo()) && nonNull(v.getVideo().getInfo()))
            v.getVideo()
                .getInfo()
                .stream()
                .filter(v -> nonNull(v.targetFile))
                .map(v -> v.targetFile.toPath())
                .forEach(p -> Try(() -> Files.deleteIfExists(p)));
    }

    @Override
    public void finishDownload() {

        try {
            Path fileWithExtension = target.resolveSibling(getDefinitiveFileName());
            Files.deleteIfExists(target);

            if (hasOnlyOneStream()) {
                target = Files.move(v.getVideo().getInfo().get(0).targetFile.toPath(), fileWithExtension, StandardCopyOption.REPLACE_EXISTING);
            } else {
                Path audioFile = getStream("audio");
                Path video = getStream("video");

                target = ffmpegService.mergeAudioAndVideo(video, audioFile, fileWithExtension);

                Files.deleteIfExists(video);
                Files.deleteIfExists(audioFile);
            }
        } catch (Exception e) {
            log.error("Error during specific move", e);
            failDownload();
            throw new RuntimeException("Error during specific move", e);
        }

        super.finishDownload();
    }

    private boolean hasOnlyOneStream() {
        return v.getVideo().getInfo().size() == 1;
    }

    private Path getStream(String type) {
        return v.getVideo().getInfo().stream()
                .filter(v -> v.getContentType().contains(type))
                .map(v -> v.targetFile.toPath())
                .findFirst()
                .orElseThrow(() -> new RuntimeException(format(ERROR_NO_CONTENT_TYPE, type, item.getTitle(), item.getUrl())));
    }

    @Override
    public Integer compatibility(DownloadingItem ditem) {
        return ditem.getUrls().length() == 1 && ditem.getUrls().head().contains("www.youtube.com")
                ? 1
                : Integer.MAX_VALUE;
    }

    private String getDefinitiveFileName() {
        String videoExt = v.getVideo().getInfo().stream()
                .map(VideoFileInfo::getContentType)
                .filter(c -> c.contains("video"))
                .map(c -> StringUtils.substringAfter(c, "/"))
                .findFirst()
                .orElse(DEFAULT_EXTENSION_MP4);

        return target.getFileName().toString().replace(temporaryExtension, "." + videoExt);
    }

    @Slf4j
    static class YoutubeWatcher implements Runnable {

        private final YoutubeDownloader youtubeDownloader;
        private final ZonedDateTime launchDateDownload = now();
        private Long globalSize = null;
        Integer MAX_WAITING_MINUTE = 5;

        public YoutubeWatcher(YoutubeDownloader youtubeDownloader) {
            this.youtubeDownloader = youtubeDownloader;
        }

        @Override
        public void run() {
            VideoInfo info = youtubeDownloader.v.getVideo();
            List<VideoFileInfo> downloadInfo = info.getInfo();
            Item item = youtubeDownloader.item;

            if (item.getStatus() == Status.PAUSED) {
                synchronized (this) { Try.run(this::wait); }
            }

            switch (info.getState()) {
                case EXTRACTING_DONE:
                    log.debug(FilenameUtils.getName(valueOf(item.getUrl())) + " " + info.getState());
                    break;
                case ERROR:
                    youtubeDownloader.failDownload();
                    break;
                case DONE:
                    downloadInfo
                            .stream()
                            .map(vi -> vi.targetFile)
                            .filter(Objects::nonNull)
                            .forEach(f -> log.debug("{} - Téléchargement terminé", FilenameUtils.getName(f.getAbsolutePath())));
                    if (item.getStatus() == Status.STARTED)
                        youtubeDownloader.finishDownload();
                    break;
                case RETRYING:
                    log.debug(info.getState() + " " + info.getDelay());
                    if (info.getDelay() == 0) {
                        log.error(info.getException().toString());
                    }
                    if (DownloadIOCodeError.class.isInstance(info.getException())) {
                        log.debug("Cause  : " + DownloadIOCodeError.class.cast(info.getException()).getCode());
                    }

                    if (launchDateDownload.isBefore(now().minusHours(MAX_WAITING_MINUTE))) {
                        youtubeDownloader.failDownload();
                    }
                    break;
                case DOWNLOADING:
                    downloading(downloadInfo, item);
                    break;
                case STOP:
                    log.debug("Pause / Arrêt du téléchargement du téléchargement");
                    break;
                default:
                    break;
            }
        }

        private void downloading(List<VideoFileInfo> downloadInfo, Item item) {

            if (isNull(globalSize)) {
                globalSize = downloadInfo.stream()
                        .filter(v -> nonNull(v.getLength()))
                        .mapToLong(URLInfo::getLength)
                        .sum();
            }

            Long count = downloadInfo.stream().mapToLong(DownloadInfo::getCount).sum();
            int currentState = (int) (count * 100 / (float) globalSize );
            if (item.getProgression() < currentState) {
                item.setProgression(currentState);
                log.debug("{} - {}%", item.getTitle(), item.getProgression());
                youtubeDownloader.convertAndSaveBroadcast();
            }
        }
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/downloader/YoutubeDownloader.kt
package com.github.davinkevin.podcastserver.manager.downloader

import com.github.axet.vget.VGet
import com.github.axet.vget.info.VideoFileInfo
import com.github.axet.vget.info.VideoInfo
import com.github.axet.wget.info.ex.DownloadIOCodeError
import com.github.axet.wget.info.ex.DownloadMultipartError
import com.github.davinkevin.podcastserver.service.FfmpegService
import com.github.davinkevin.podcastserver.service.MimeTypeService
import com.github.davinkevin.podcastserver.service.factory.WGetFactory
import com.github.davinkevin.podcastserver.service.properties.PodcastServerParameters
import io.vavr.API.Try
import lan.dk.podcastserver.entity.Item
import lan.dk.podcastserver.entity.Status
import lan.dk.podcastserver.repository.ItemRepository
import lan.dk.podcastserver.repository.PodcastRepository
import org.apache.commons.io.FilenameUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_PROTOTYPE
import org.springframework.context.annotation.Scope
import org.springframework.messaging.simp.SimpMessagingTemplate
import org.springframework.stereotype.Component
import java.lang.String.format
import java.net.URL
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.time.Duration
import java.time.ZonedDateTime.now
import java.time.temporal.TemporalAmount
import java.util.Objects.nonNull
import java.util.concurrent.CountDownLatch

/**
 * Created by kevin on 14/12/2013 for Podcast Server
 */

@Component
@Scope(SCOPE_PROTOTYPE)
class YoutubeDownloader(
        itemRepository: ItemRepository,
        podcastRepository: PodcastRepository,
        podcastServerParameters: PodcastServerParameters,
        template: SimpMessagingTemplate,
        mimeTypeService: MimeTypeService,
        val wGetFactory: WGetFactory,
        val ffmpegService: FfmpegService
) : AbstractDownloader(itemRepository, podcastRepository, podcastServerParameters, template, mimeTypeService) {

    private val log = LoggerFactory.getLogger(YoutubeDownloader::class.java)

    lateinit var v: VGet

    private lateinit var watcher: YoutubeWatcher

    override fun download(): Item {
        try {
            watcher = YoutubeWatcher(this)
            val parser = wGetFactory.parser(item.url)

            v = wGetFactory.newVGet(parser.info(URL(item.url)))
            v.extract(parser, stopDownloading, watcher)

            target = getTargetFile(item)

            v
                    .video
                    .info
                    .forEach { vi -> vi.target = generatePartFile(target!!, vi).toFile() }

            v.download(parser, stopDownloading, watcher)

        } catch (e: Exception) {
            if (e is DownloadMultipartError) {
                e.info
                        .parts
                        .stream()
                        .map { it.exception }
                        .filter { it != null }
                        .forEach { it.printStackTrace() }
            }
            throw e
        }

        log.debug("Download ended")
        return item
    }

    private fun generatePartFile(targetFile: Path, vi: VideoFileInfo): Path {
        return targetFile.resolveSibling(targetFile.fileName.toString() + v.getContentExt(vi))
    }

    override fun getFileName(item: Item) =
            v.video?.title?.replace("[^a-zA-Z0-9.-]".toRegex(), "_") ?:
            throw RuntimeException("Error during creation of filename of " + item.url)

    override fun pauseDownload() {
        item.status = Status.PAUSED
        saveSyncWithPodcast()
        convertAndSaveBroadcast()
    }

    override fun restartDownload() {
        item.status = Status.STARTED
        saveSyncWithPodcast()
        convertAndSaveBroadcast()
        watcher.lock.countDown()
    }

    override fun stopDownload() {
        if (item.status == Status.PAUSED) {
            watcher.lock.countDown()
        }
        super.stopDownload()

        if (nonNull(v) && nonNull(v.video) && nonNull(v.video.info))
            v.video
                    .info
                    .stream()
                    .filter { v -> nonNull(v.targetFile) }
                    .map { v -> v.targetFile.toPath() }
                    .forEach { p -> Try { Files.deleteIfExists(p) } }
    }

    override fun finishDownload() {
        val fileWithExtension = target!!.resolveSibling(toDefinitiveFileName())
        Files.deleteIfExists(target)

        target =
                if (hasOnlyOneStream()) {
                    try { Files.move(v.video.info[0].targetFile.toPath(), fileWithExtension, StandardCopyOption.REPLACE_EXISTING) }
                    catch (e: Exception) { failDownload(); throw RuntimeException("Error during specific move", e) }
                } else {
                    val audioFile = getStream("audio")
                    val video = getStream("video")
                    try { ffmpegService.mergeAudioAndVideo(video, audioFile, fileWithExtension)
                    } catch (e: Exception) { failDownload(); throw RuntimeException("Error during specific move", e)
                    } finally {
                        Files.deleteIfExists(video)
                        Files.deleteIfExists(audioFile)
                    }
                }
        super.finishDownload()
    }

    private fun toDefinitiveFileName(): String {
        val videoExt = v.video.info.asSequence()
                .map { it.contentType }
                .filter { it.contains("video") }
                .map { it.substringAfter("/") }
                .firstOrNull() ?: DEFAULT_EXTENSION_MP4

        return target!!.fileName.toString().replace(temporaryExtension, ".$videoExt")
    }

    private fun hasOnlyOneStream() = v.video.info.size == 1

    private fun getStream(type: String) =
            v.video.info.asSequence()
                    .filter { v -> v.contentType.contains(type) }
                    .map { v -> v.targetFile.toPath() }
                    .firstOrNull() ?: throw RuntimeException(format(ERROR_NO_CONTENT_TYPE, type, item.title, item.url))

    override fun compatibility(downloadingItem: DownloadingItem) =
            if (downloadingItem.urls.size == 1 && "youtube.com" in downloadingItem.urls.first()) 1
            else Integer.MAX_VALUE

    internal class YoutubeWatcher(private val youtubeDownloader: YoutubeDownloader, private val maxWaitingTime: TemporalAmount = Duration.ofMinutes(5)) : Runnable {

        private val log = LoggerFactory.getLogger(YoutubeWatcher::class.java)
        var lock = CountDownLatch(1)

        private val launchDateDownload = now()
        private var globalSize: Long = -1L

        override fun run() {
            val info = youtubeDownloader.v.video
            val downloadInfo = info.info
            val item = youtubeDownloader.item

            when (info.state) {
                VideoInfo.States.EXTRACTING_DONE -> log.debug(FilenameUtils.getName(item.url.toString()) + " " + info.state)
                VideoInfo.States.ERROR -> youtubeDownloader.failDownload()
                VideoInfo.States.DONE -> done(downloadInfo, item)
                VideoInfo.States.RETRYING -> retrying(info)
                VideoInfo.States.DOWNLOADING -> downloading(downloadInfo, item)
                VideoInfo.States.STOP -> log.debug("Pause / Stop download")
                else -> {}
            }

            if (item.status == Status.PAUSED) {
                awaitAndRestart()
            }
        }

        private fun retrying(info: VideoInfo) {
            val exception = info.exception

            log.debug(info.state.toString() + " " + info.delay)
            if (info.delay == 0) {
                log.error(exception.toString())
            }
            if (exception is DownloadIOCodeError) {
                log.debug("Cause  : " + exception.code)
            }

            if (launchDateDownload.isBefore(now().minus(maxWaitingTime))) {
                youtubeDownloader.failDownload()
            }
        }

        private fun downloading(downloadInfo: List<VideoFileInfo>, item: Item) {
            if (globalSize == -1L) {
                globalSize = downloadInfo.filter { it.length != null }.map { it.length }.sum()
            }

            val count = downloadInfo.map { it.count }.sum()
            val currentState = (count * 100 / globalSize.toFloat()).toInt()
            if (item.progression < currentState) {
                item.progression = currentState
                log.debug("{} - {}%", item.title, item.progression)
                youtubeDownloader.convertAndSaveBroadcast()
            }
        }

        private fun done(downloadInfo: List<VideoFileInfo>, item: Item) {
            downloadInfo
                    .stream()
                    .map { it.targetFile }
                    .filter { it != null }
                    .forEach { log.debug("{} - Téléchargement terminé", FilenameUtils.getName(it.absolutePath)) }

            if (item.status == Status.STARTED)
                youtubeDownloader.finishDownload()
        }

        private fun awaitAndRestart() {
            lock.await()
            lock = CountDownLatch(1)
        }
    }

    companion object {
        private const val DEFAULT_EXTENSION_MP4 = "mp4"
        private const val ERROR_NO_CONTENT_TYPE = "Content Type %s not found for video %s at url %s"
    }
}
